### DISCUSSION TITLE
Why use move_uploaded_file in the example of sending multiple files?

### DISCUSSION THREAD

**Discussion by accountnujen** [PHPMailer/PHPMailer#2803] [Q&A]
You have a sample file that contains code like this:
```
...
    if (move_uploaded_file($_FILES['userfile']['tmp_name'][$ct], $uploadfile)) {
        if (!$mail->addAttachment($uploadfile, $filename)) {
            $msg .= 'Failed to attach file ' . $filename;
        }
    } else {
        $msg .= 'Failed to move file to ' . $uploadfile;
    }
...
```
Can you please tell me why you need to check move_uploaded_file? Why can't you just do it like this:
```
...
        if (!$mail->addAttachment($file['tmp_name'][$i], $file['name'][$i])) {
            $msg .= 'Failed to attach file ' . $filename;
        }
...
```
Can you give an example where move_uploaded_file returns false and $mail->addAttachment() returns true?

**Comment by Synchro** [comment]
Have a read of [the PHP docs on handling uploads](https://www.php.net/manual/en/features.file-upload.post-method.php). Essentially, you can't trust what is in the `$_FILES` superglobal, and you must validate it before trusting its content, and that's what [`move_uploaded_file`](https://www.php.net/manual/en/function.is-uploaded-file.php](https://www.php.net/manual/en/function.move-uploaded-file.php)) and [`is_uploaded_file`](https://www.php.net/manual/en/function.is-uploaded-file.php) do.

**Reply by accountnujen** [reply]
What could be a file that you shouldn't trust? It's not like I'm checking the contents of files or initializing a file check with antivirus. Even if the file is malicious, how does renaming the file help me?

**Reply by Synchro** [reply]
It's not concerned about the contents of the file (that's a separate concern), but where the information about things like its filename and other metadata came from. For example if you were able to inject data into `$_FILES`, you could conceivably trick a form into sending you the contents of some sensitive file on your system instead of one that was uploaded. [More examples here](https://www.prplbx.com/resources/blog/exploiting-file-upload-vulnerabilities/).

**Reply by accountnujen** [reply]
I read this article. It talks about files that are written to disk. In our case, we are dealing with temporary files, the purpose of which is to be mail files. They don't need to be saved to disk. Then what is the point of this action? Again, overwriting a file does not check its extension or give it a new name. If the mail server misses it, you can send the .htaccess file without any problems. Then what exactly are we protecting ourselves against?

**Reply by Synchro** [reply]
You seem determined to understate the potential for abuse here. A temp file is still a file on disk like any other (by the time you receive it in PHP, it's already written to disk), and if you don't validate this user-controlled input, anyone able to manipulate the contents of `$_FILES` can trick your script into either overwriting a local file or sending a sensitive file by email. Using your example of `.htaccess`, this could allow an attacker to download password hashes contained in it (to crack offline), or substitute their own, giving them unauthorised access. If you don't regard those as vulnerabilities I don't really know what to say.

Your particular script may not be subject to these issues, but in general it's something you should do anyway. Explaining why a particular situation might not be subject to this exposure would be far more involved than simply coding defensively as a matter of course. Common advice in the pentests I'm involved in is to treat everything with suspicion. Context is significant: your current code might not be vulnerable, but future code that you (or someone else) writes that uses it in a different way might change that.
